server.go

```diff
@@ -407,25 +517,70 @@ func (h *handler) runStream(
 		}
 	}
 	rw := internal.NewFnReadWriter(r, func(ctx context.Context, r *goatorepo.Rpc) error {
-		h.writeChan <- r
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case h.writeChan <- r:
+			break
+		}
 		return nil
 	})
 
 	defer handler.cancel()
 
+	var appErr error
 	si := &grpc.StreamServerInfo{
 		FullMethod:     fmt.Sprintf("/%s/%s", info.name, sd.StreamName),
 		IsClientStream: sd.ClientStreams,
 		IsServerStream: sd.ServerStreams,
 	}
 
+	var statsBegin *stats.Begin
+	incomingMetadata, imOk := metadata.FromIncomingContext(ctx)
+	if !imOk {
+		incomingMetadata = metadata.MD{}
+	}
+	for _, sh := range h.srv.statsHandlers {
+		ctx = sh.TagRPC(ctx, &stats.RPCTagInfo{
```

Comment:
This block of code also looks used often. Is this worth having as a util too? The same lines in the `InHeader` struct are commented out in the last 3 instances.